pkgerrors.go

```diff
@@ -8,50 +8,112 @@ import (
 
 type pkgError struct {
 	Err        error
-	Message    string
+	Messages   []string
 	StackTrace StackTrace
 }
 
 const (
 	pkgErrorsMessageDelimiter = ": "
 )
 
+// convertPkgError converts pkg/errors to fail.
+// It returns nil if the error is not derived from pkg/errors.
+func convertPkgError(err error) (convertedErr *Error) {
+	pkgErr := extractPkgError(err)
+	if pkgErr == nil {
+		return
+	}
+
+	if appErr, ok := pkgErr.Err.(*Error); ok {
+		convertedErr = appErr.Copy()
+		convertedErr.StackTrace = mergeStackTraces(appErr.StackTrace, pkgErr.StackTrace)
+	} else {
+		convertedErr = &Error{
+			Err:        pkgErr.Err,
+			StackTrace: pkgErr.StackTrace,
+		}
+	}
+
+	for i := len(pkgErr.Messages) - 1; i >= 0; i-- {
+		WithMessage(pkgErr.Messages[i])(convertedErr)
+	}
+
+	return
+}
+
 // extractPkgError extracts the innermost error from the given error.
 // It converts the stack trace that is annotated by pkg/errors into fail.StackTrace.
 // If the error doesn't have a stack trace or a causer of pkg/errors,
 // it simply returns the original error
-func extractPkgError(err error) pkgError {
+func extractPkgError(err error) *pkgError {
 	type traceable interface {
 		StackTrace() pkgerrors.StackTrace
 	}
 	type causer interface {
 		Cause() error
 	}
 
-	rootErr := err
 	var stackTraces []StackTrace
+	var messages []string
+	var lastMessage string
+
+	// Retrive stacks and trace back the root cause
+	rootErr := err
 	for {
 		if t, ok := rootErr.(traceable); ok {
 			stackTrace := convertStackTrace(t.StackTrace())
 			stackTraces = append(stackTraces, stackTrace)
 		}
 
 		if cause, ok := rootErr.(causer); ok {
+			msg := rootErr.Error()
+			if lastMessage != msg {
+				lastMessage = msg
+				messages = append(messages, msg)
+			}
+
 			rootErr = cause.Cause()
 			continue
 		}
 
 		break
 	}
 
-	var msg string
-	if strings.HasSuffix(err.Error(), pkgErrorsMessageDelimiter+rootErr.Error()) {
-		msg = strings.TrimSuffix(err.Error(), pkgErrorsMessageDelimiter+rootErr.Error())
+	if len(stackTraces) == 0 && rootErr == err {
```

Comment:
```go
if len(stackTraces) == 0 && rootErr == err {
   ^^^^^^^^^^^^^^^^^^^^^    ^^^^^^^^^^^^^^
   1                        2
```

1. Checks conformance to `traceable` interface.
2. Checks conformance to `causer` interface.

---

https://github.com/pkg/errors/blob/27936f6d90f9c8e1145f11ed52ffffbfdb9e0af7/errors.go

`New` and `Errorf` use `&fundamental{}`.

```go
// fundamental is an error that has a message and a stack, but no caller.
type fundamental struct {
	msg string
	*stack
}
```

`WithStack` uses `&withStack{}`.

```go
type withStack struct {
	error
	*stack
}
```

`Wrap` and `Wrapf` use `&withStack{&withMessage{}}`;
`WithMessage` and `WithMessagef` use `&withMessage{}` alone.

```go
type withMessage struct {
	cause error
	msg   string
}
```

---

`*stack` conforms to `traceable` interface.

https://github.com/pkg/errors/blob/27936f6d90f9c8e1145f11ed52ffffbfdb9e0af7/stack.go#L155-L161

`*withStack` conforms to both `traceable` and `causer` interfaces.

https://github.com/pkg/errors/blob/27936f6d90f9c8e1145f11ed52ffffbfdb9e0af7/errors.go#L160

`*withMessage` conforms to `causer` interface.